### ISSUE TITLE
Multi-task finetuning with regression and classification

### ISSUE THREAD

**Issue by Andrew-Wyn** [Andrew-Wyn/augmenting_nlms_meco#4]

**Comment by lucadinidue** [comment]
**Multi Class Bert (just classification - no regression)**

    from transformers import BertTokenizer, BertForTokenClassification, BertConfig
    import torch.nn as nn
    
    class MultiClassBERT(nn.Module):
        def __init__(self, model_name, num_labels_list):
            super(MultiClassBERT, self).__init__()
            
            self.bert = BertForTokenClassification.from_pretrained(model_name, num_labels=1, return_dict=True)
            
            self.classifiers = nn.ModuleList([
                nn.Linear(self.bert.config.hidden_size, num_labels) for num_labels in num_labels_list
            ])
        
        def forward(self, input_ids, attention_mask):
            outputs = self.bert(input_ids, attention_mask=attention_mask)
            logits_list = [classifier(outputs.last_hidden_state) for classifier in self.classifiers]
            
            return logits_list
    
    # Example of how to initialize and train the model
    tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
    model = MultiClassBERT("bert-base-uncased", [3, 4, 5])  # Three class sets with 3, 4, and 5 classes respectively
    
    # Example input
    text = "Hello world"
    tokens = tokenizer.tokenize(text)
    input_ids = torch.tensor(tokenizer.encode(text)).unsqueeze(0)  # Batch size 1
    
    # Get the logits for each class set
    logits_list = model(input_ids, attention_mask=(input_ids != tokenizer.pad_token_id))
    
    # Compute the loss for each class set using the appropriate labels for each class set
    # loss = sum of all individual losses

**Multi Class Data Collator**
    
    from transformers import DataCollatorForTokenClassification, PreTrainedTokenizer
    import torch
    
    class MultiLabelDataCollator(DataCollatorForTokenClassification):
        def __init__(self, tokenizer: PreTrainedTokenizer, pad_to_multiple_of: Optional[int] = None):
            super().__init__(tokenizer, pad_to_multiple_of)
    
        def __call__(self, features):
            # Separate labels from features
            all_label_sets = list(zip(*[feature['labels'] for feature in features]))
            
            # Use the parent class's collate function to handle tokens, input_ids, etc.
            batch = super().__call__(features)
    
            # Adjust the 'labels' key to handle multiple sets of labels
            batch['labels'] = [torch.nn.functional.pad(torch.tensor(label_set, dtype=torch.int64), 
                          (0, self.tokenizer.model_max_length - len(label_set)), 
                          value=self.tokenizer.pad_token_id) for label_set in all_label_sets]
    
            return batch

**Comment by lucadinidue** [comment]
- Ho controllato che [gilberto](https://huggingface.co/idb-ita/gilberto-uncased-from-camembert) effettivamente caricasse CamembertModel ed è così.
- Ho caricato anche Camembert Multitask ma ho visto che il codice è copiato da RoBERTa, quindi ho deciso di non stare a implementare anche XLM-RoBERTa perchè pure quello è copiato  
`# Copied from transformers.models.roberta.modeling_roberta.RobertaForTokenClassification with Roberta->XLMRoberta, ROBERTA->XLM_ROBERTA`
quindi aspetterei di avere la versione definitiva di RoBERTa
